sw/device/lib/dif/dif_rv_timer.h

```diff
@@ -0,0 +1,321 @@
+ * @param threshold The value to go off at.
+ * @return The result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_arm(const dif_rv_timer_t *timer,
+                                       uint32_t hart_id, uint32_t comp_id,
+                                       uint64_t threshold);
+
+/**
+ * Enables or disables a particular timer's IRQ, such that when the threshold is
+ * reached, no interrupt is fired.
+ *
+ * @param timer A timer device.
+ * @param hart_id The hart counter associated with the timer.
+ * @param comp_id The comparator associated with the timer.
+ * @param state The desired status.
+ * @return the result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_irq_enable(const dif_rv_timer_t *timer,
+                                              uint32_t hart_id,
+                                              uint32_t comp_id,
+                                              dif_rv_timer_enabled_t state);
+
+/**
+ * Returns whether the IRQ for a particular timer is currently being serviced.
+ *
+ * @param timer A timer device.
+ * @param hart_id The hart counter associated with the timer.
+ * @param comp_id The comparator associated with the timer.
+ * @param flag_out Out-param for the interrupt status.
+ * @return the result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_irq_get(const dif_rv_timer_t *timer,
+                                           uint32_t hart_id, uint32_t comp_id,
+                                           bool *flag_out);
+
+/**
+ * Clears the IRQ for a particular timer.
+ *
+ * @param timer A timer device.
+ * @param hart_id The hart counter associated with the timer.
+ * @param comp_id The comparator associated with the timer.
+ * @return the result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_irq_clear(const dif_rv_timer_t *timer,
+                                             uint32_t hart_id,
+                                             uint32_t comp_id);
+
+/**
+ * Forces the IRQ for a particular timer to fire.
+ *
+ * @param timer A timer device.
+ * @param hart_id The hart counter associated with the timer.
+ * @param comp_id The comparator associated with the timer.
+ * @return the result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_irq_force(const dif_rv_timer_t *timer,
+                                             uint32_t hart_id,
+                                             uint32_t comp_id);
+
+/**
+ * Disables all IRQs for a particular hart.
+ *
+ * Optionally returns a `state` value containing the previous itnerrupt state.
+ * `state` may be NULL. See `dif_rv_timer_irq_restore()`.
+ *
+ * @param timer a timer device.
+ * @param state out param for IRQ state information.
+ * @return the result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_irq_disable(const dif_rv_timer_t *timer,
+                                               uint32_t hart_id,
+                                               uint32_t *state);
```

Comment:
It is pretty confusing to me that `dif_rv_timer_irq_enable()` is used to enable or disable interrupts for a specific hart+comparator combo, while `dif_rv_timer_irq_disable()` is used to disable all IRQs for a given hart -- from the naming I would expect them to do basically the same thing, with one enabling and one disabling.

It seems like `dif_rv_timer_irq_enable()` should be named something like `dif_rv_timer_irq_config_enable()`, and this should be named something like `dif_rv_timer_irq_disable_all()` or something.

Reply:
I agree with @hudson-ayers , I don't think this particular comment is "bikeshedding", it is pretty confusing. I would say that it would be a good idea to change `dif_rv_timer_irq_disable` to `dif_rv_timer_irqs_disable` and the same with restore.

Reply:
I think this discussion falls under this comment/issue: https://github.com/lowRISC/opentitan/issues/2366#issuecomment-637071965 

I really don't want to hold up this review when the design review guidance has been "match the UART API". Let's take naming discussions to the linked issue.

Reply:
Sounds good! Sorry about that.